Wikipedia/Code/DiffContainerViewController.swift

```diff
@@ -1148,43 +1187,118 @@ extension DiffContainerViewController: DiffToolbarViewDelegate {
+        let undo = UIAlertAction(title: CommonStrings.undo, style: .destructive) { [weak self] (action) in
+            self?.performUndo()
+        }
         undo.isEnabled = false
-
-        // DIFFTODO: Hook up undo action and enable only if textfield is non-empty
+        undoAlertUndoAction = undo
 
         alertController.addAction(cancel)
         alertController.addAction(undo)
 
         present(alertController, animated: true)
     }
+    
+    @objc private func undoSummaryTextfieldDidChange() {
+        undoAlertUndoAction?.isEnabled = (undoAlertSummaryTextField?.text?.count ?? 0) > 0
+    }
+    
+    private func performUndo() {
+        guard let wkProject = wkProject,
+              let title = articleTitle,
+        let revisionID = toModelRevisionID,
+        let username = toModel?.user,
+        let summary = undoAlertSummaryTextField?.text else {
+            return
+        }
+        
+        fakeProgressController.start()
+        WKWatchlistService().undo(title: title, revisionID: UInt(revisionID), summary: summary, username: username, project: wkProject) { [weak self] result in
+
+            DispatchQueue.main.async {
+                self?.completeRollbackOrUndo(result: result, isRollback: false)
+            }
+        }
+    }
 
     func tappedRollback() {
         let title = WMFLocalizedString("diff-rollback-alert-title", value: "Rollback edits", comment: "Title of alert when user taps rollback in diff toolbar.")
         let message = WMFLocalizedString("diff-rollback-alert-message", value: "Are you sure you want to rollback the edits?", comment: "Message in alert when user taps rollback in diff toolbar.")
         let alertController = UIAlertController(title: title, message: message, preferredStyle: .alert)
 
         let cancel = UIAlertAction(title: CommonStrings.cancelActionTitle, style: .cancel)
-        let rollback = UIAlertAction(title: CommonStrings.rollback, style: .destructive)
-
-        // DIFFTODO: Hook up rollback action
-
+        let rollback = UIAlertAction(title: CommonStrings.rollback, style: .destructive) { [weak self] (action) in
+            self?.performRollback()
+        }
+        
         alertController.addAction(cancel)
         alertController.addAction(rollback)
 
         present(alertController, animated: true)
     }
-
+    
+    private func performRollback() {
+        guard let wkProject = wkProject,
+              let title = articleTitle,
+              let username = toModel?.user else {
+            return
+        }
+        
+        fakeProgressController.start()
+        WKWatchlistService().rollback(title: title, project: wkProject, username: username) { [weak self] result in
+            DispatchQueue.main.async {
+                self?.completeRollbackOrUndo(result: result, isRollback: true)
+            }
+        }
+    }
+    
+    private func completeRollbackOrUndo(result: Result<WKUndoOrRollbackResult, Error>, isRollback: Bool) {
```

Comment:
Let's **not** make any changes here (because I believe I understand the tradeoffs made in this case and it's working well as-is 😀), but just wanted to make a suggestion for thinking about future API design decisions.

When reviewing, this declaration stood out to me as slightly confusing:
`private func completeRollbackOrUndo(result: Result<WKUndoOrRollbackResult, Error>, isRollback: Bool)`

and as something that could be simplified and made more sustainable as something like:
`private func completeRevert(result: Result<WKRevertResult, Error>)`

My recommendation would be that the `enum` type would be more suited than `bool` as an underlying type indicator in cases like this. `Bool` is certainly useful in some cases when we're sure there are and will only be two discrete states like on/off or enabled/disabled, but even though there are _currently_ only two options in this case, `isRollback = false` doesn't immediately and precisely tell me that it's an undo the same way `isUndo = false` wouldn't precisely tell me that the action is a rollback. I personally dunno if "revert" is exactly the right grouping word for these actions or not, but I think having some parent type to represent these actions is more sustainable than using the `Bool` type here.

A fuller `WKData`/client app sketch:

```swift
struct WKRevertResult {
	let action: WKRevertAction
	let newRevisionID: Int
	let oldRevisionID: Int
}

enum WKRevertAction {
	case rollback
	case undo
}

public func rollback(title: String, project: WKProject, username: String, completion: @escaping (Result<WKRevertResult, Error>) -> Void) {
	...
	completion(.success(WKRevertResult(action: .rollback, newRevisionID: newRevisionID, oldRevisionID: oldRevisionID)))
}

public func undo(title: String, revisionID: UInt, summary: String, username: String, project: WKProject, completion: @escaping (Result<WKUndoOrRollbackResult, Error>) -> Void) {
	...
	completion(.success(WKRevertResult(action: .undo, newRevisionID: newRevisionID, oldRevisionID: oldRevisionID)))
}


private func completeRevert(result: Result<WKRevertResult, Error>) {
	...
}
```

Totally open to different perspectives here, but to me benefits of the modified approach:

1. if our "revert" action types grow beyond just undo and rollback, our method signature doesn't need to change
2. all the interesting details (the action itself and the revisions we're interested in) stay contained in the result and don't leak out

Reply:
@staykids thanks for the write-up! I definitely agree, your approach does feel more clean. Leaving as-is per your request, but I'll keep an eye out for this habit of mine in the future.